sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala

```diff
@@ -323,12 +316,12 @@ private[sql] object SparkSqlSerializer2 {
    */
   def createDeserializationFunction(
       schema: Array[DataType],
-      in: DataInputStream,
-      mutableRow: SpecificMutableRow): () => Unit = {
+      in: DataInputStream): () => Row = {
     () => {
       // If the schema is null, the returned function does nothing when it get called.
       if (schema != null) {
         var i = 0
+        val mutableRow = new GenericMutableRow(schema.length)
```

Comment:
@yhuai and I chatted about this offline.  The reason that we need to perform this copy is because this patch allows SqlSerializer2 to be used in cases where the shuffle performs a sort.  In HashShuffleReader, Spark ends up passing the iterator returned from this deserializer to ExternalSorter, which buffers rows because it needs to sort them based on their contents.

I think that we only need to copy the row in cases where we're shuffling with a key ordering.  To avoid unnecessary copying in other cases, I think that we can extend `SparkSqlSerializer2`'s constructor to accept a boolean flag that indicates whether we should copy, and should thread that flag all the way down to here.  In `Exchange`, where we create the serializer, we can check whether the shuffle will use a keyOrdering; if it does, then we'll enable copying.  Avoiding this copy in other cases should provide a nice performance boost for aggregation queries.